src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -1081,6 +1082,73 @@ private void ApplyTextDocumentChange(DocumentId documentId, SourceText newText)
             }
         }
 
+        protected override void ApplyDocumentInfoChanged(DocumentId id, DocumentInfo updatedInfo)
+        {
+            var document = CurrentSolution.GetDocument(id);
+
+            FailIfDocumentInfoChangesNotSupported(document, updatedInfo);
+
+            if (document.Name != updatedInfo.Name)
+            {
+                GetProjectData(updatedInfo.Id.ProjectId, out var _, out var project);
+                var projectItemForDocument = project.FindItem(document.Name, StringComparer.InvariantCultureIgnoreCase); //?
+
+                if (projectItemForDocument == null)
+                {
+                    // TODO: The document might actually belong to a Shared Project.
+                    return;
+                }
+
+                // Must save the document first for things like Breakpoints to be preserved.
+                projectItemForDocument.Save();
+
+                var uniqueName = projectItemForDocument.Collection.GetUniqueName(
+                    Path.GetFileNameWithoutExtension(updatedInfo.Name),
+                    Path.GetExtension(updatedInfo.Name));
+
+                // By setting this property, Visual Studio will perform the file rename, which 
+                // will cause the workspace's current solution to update and will fire the 
+                // necessary workspace changed events.
+                projectItemForDocument.Name = uniqueName;
```

Comment:
@jmarolf Can you take a look at this part and confirm it looks good to you?

Reply:
This will synchronously block the UI thread on legacy.  On the new stuf it will queue an operation in the background.  As long as you don't immediately expect the workspace to be populated with these changes (rename happens after all other workspace changes) then this is fine.

---
In reply to: [265312543](https://github.com/dotnet/roslyn/pull/33733#discussion_r265312543) [](ancestors = 265312543)

Reply:
Did some testing locally this looks great